haystack/core/pipeline/pipeline.py

```diff
@@ -390,6 +393,34 @@ def run(  # noqa: PLR0915, PLR0912, C901, pylint: disable=too-many-branches
                 except PipelineRuntimeError as error:
                     # Attach partial pipeline outputs to the error before re-raising
                     error.pipeline_outputs = pipeline_outputs
+                    # Create a snapshot of the last good state of the pipeline before the error occurred.
+                    pipeline_snapshot_inputs_serialised = deepcopy(inputs)
+                    pipeline_snapshot_inputs_serialised[component_name] = deepcopy(component_inputs)
+                    out_dir = get_output_dir("pipeline_snapshot")
+                    last_good_state_snapshot = _create_pipeline_snapshot(
+                        inputs=pipeline_snapshot_inputs_serialised,
+                        # Dummy breakpoint to pass the component_name and state_persistence_path to the
+                        # _save_pipeline_snapshot
+                        break_point=Breakpoint(
+                            component_name=component_name, visit_count=0, snapshot_file_path=out_dir
+                        ),
+                        component_visits=component_visits,
+                        original_input_data=data,
+                        ordered_component_names=ordered_component_names,
+                        include_outputs_from=include_outputs_from,
+                        pipeline_outputs=pipeline_outputs,
+                    )
+                    f_name = f"last_good_state_{datetime.now().strftime('%Y_%m_%d_%H_%M_%S')}.json"
+                    try:
+                        _save_pipeline_snapshot(pipeline_snapshot=last_good_state_snapshot, file_name=f_name)
+                    except Exception:
+                        msg = "Failed to save a snapshot of the pipeline's last valid state to '{f_name}'."
+                        logger.error(msg)
+                    logger.info(
+                        "Saved a snapshot of the pipeline's last valid state to '{f_name}'. "
+                        "Review this snapshot to debug the error and resume the pipeline from here.",
+                        f_name=f_name,
+                    )
```

Comment:
This should be moved to be right under `_save_pipeline_snapshot` within the `try` otherwise we could send both the logger.error and logger.info which doesn't make sense.